Added Identity documentation (#302)
microstates/microstates.js: README.md

</ul>;
```

# Streaming State

> A microstate represents a single immutable value and provides transitions to derive the next value. Microstates provides a mechanism called Identity that allows to emit stream of Microstates. When you create an Identity from a Microstate, you get an object that has the same shape the original microstate. Every composed microstate becomes an identity and every transition gets wrapped in side effect emitting behaviour specific to the identity's constructor. Identity wrapped Microstate offer the following benefits over raw Microstates.

Reviewer: ```suggestion
A microstate represents a single immutable value with transitions to derive the next value. Through a mechanism called `Identity`, you can emit stream of Microstates as the value is updated over time.

When you create an Identity from a Microstate, you get an object that has the same shape the original microstate. Every composed microstate becomes an identity and every transition gets wrapped in side effect emitting behavior specific to the identity's constructor. Identity wrapped Microstate offer the following benefits over raw Microstates.
```

> A microstate represents a single immutable value and provides transitions to derive the next value. Microstates provides a mechanism called Identity that allows to emit stream of Microstates. When you create an Identity from a Microstate, you get an object that has the same shape the original microstate. Every composed microstate becomes an identity and every transition gets wrapped in side effect emitting behaviour specific to the identity's constructor. Identity wrapped Microstate offer the following benefits over raw Microstates.

Reviewer: This seems _way_ more clear to me now. Basically, `Identity` is the original `Microstate` plus the additional logic that allows you to subscribe to the changes over time, either through `Store` or `Observable`. Is that right?

Author: Yes

</ul>;
```

# Streaming State

A microstate represents a single immutable value and provides transitions to derive the next value. Microstates may be considered pure function of type and value. In addition, Microstate transitions are pure, meaning they have no side-effects or, in other words, they do not affect anything outside of themselves. These are powerful constraints that provide us with guarantees that we can leverage to build efficient reactive systems.

To bring a reactive system to life, we need to capture the latest state in a sequence of states. In component driven reactive engines like React, Vue, Ember & Angular the state needs to emerge at the root of the component tree. When a new state is emitted, we need to apply the state to the root component and allow the state to cascade down the component tree.

> Microstates has a mechanism to working with Microtates in reactive environments. It is called Identity. When you create an Identity from a Microstate, you get an object that has the same shape the original microstate. Every composed microstate becomes an identity and every transition gets wrapped in side effect emitting behaviour specific to that identity constructor. Let’s look at the features of the Identity and how they benefit reactive environments.

Reviewer: ```suggestion
Microstates has a mechanism for working with reactive environments: Identity. When you create an Identity from a Microstate, you get an object that has the same shape the original microstate. Let’s look at the features of the Identity and how they benefit reactive environments.
```

I think there was a little too much context given up front there. Better to cut it down a bit and let the example explain how it's used.

## Structural Sharing

A common performance optimization used by all reactive engines is to prevent re-renders for components who’s props have not changed. The most efficient way to determine if a value has not changed it to perform an exact equality check, for example: `prevValue === currentValue`. If the reference is the same, then consider the value unchanged. The Identity makes this possible with Microstates by internally managing how the Identity is constructed as a result of a transition. It will automatically determine which branches of microstates are unchanged and reuse previous identities for those branches.

## Memoized Getters

Microstates are immutable which makes it safe for us to memoize computations that are derived off their state. Identies will automatically memoize getters and return previously computed value when the microstate backing the identity has not changed. When a transition is invoked on the identity, the part of the identity tree that are changed will be re-created effectively invalidating the cache for the changed parts of the identity. The getters will be recomputed for state that is changed.

> ## Debounce no-op transitions

Reviewer: How is a no-op calculated? By which I mean to ask, a transition from and back to the original state within what time period avoids emitting a new Identity? The following code performs two transitions

```javascript
let id = Store(from({ foo: 'bar' }), prev => {
  console.count('changed');
  id = prev;
});

id.foo.set('bax');
id.foo.set('bar');
```

Author: > How is a no-op calculated?

It's based on value equality. Using your example,

```js
id.foo.set('bax');
id.foo.set('bax');
```

Would only cause one update because 2nd operation is debounced.

Reviewer: Oh okay, I see what you're saying -- if the current state and the state being transitioned to are the same, then a "new state" that would be the same _is not_ emitted.

Author: Yes.

Identities automatically prevent unnecessary re-renders by debouncing transitions that do not change the value. This eliminates the need for `shouldComponentUpdate` hooks for pure components because it is safe to assume that a component that is re-rendering is re-rendering as a result of a transition that changed the value.

## Identity Constructors

> Microstates comes with two Identity constructors: _Store_ and _Observable_. Observable will create a stream of identities. Next identity will be sent to the observer when an transition on an identity is called.

Reviewer: Just for symmetry, if you're going to give a basic explanation of what `Observable` does in this instance, I think it's also worth giving a short one-sentence description of what `Store` does as well.

### Store(microstate, callback)

Store identity constructor takes two arguments: microstate and a callback. It returns an identity. When a transition is invoked on the identity, the callback will receive the next identity.

```js
import { Store, from, valueOf } from 'microstates';

let initial = create(Number, 42);

let last;

> last = Store(initial, next => (last = next));

Reviewer: Does the `Store` callback get invoked for the initialize transition? Or just every future transition?

Reviewer: To answer my own question, it seems like it's only for future transitions, so you need to use the return value from `Store` for the initial state of the `last` variable. The following _does not_ work:

```
let last;

Store(initial, next => (last = next));

last.foo.set('bax');
```

Author: Yes, initialize transition lazily sets up the microstate on create. When you construct the Microstate, the value is not initialized until you still pulling values off the microstate. If your microstate is bound to the DOM, the render will pull the values off the microstate which will invoke initialize but it's a pure operation without side effects.

last.increment();
//> undefined
// callback will be invoked syncronously on transition

// last here will reference the last
last.increment();
//> undefined

valueOf(last);
//> 44
```

The same mechanism can be used with React or any other reactive environment.

```js
import React from 'react';
import { Store, create } from 'microstates';

class Counter extends React.Component {
  state = {
    last: Store(create(Number, 42), next => this.setState({ last: next }))
  };
  render() {
    let { last } = this.state;
    return (
      <button onClick={() => last.increment()}>Increment {last.state}</button>
    );
  }
}
```

> ### Observable.from(microstate)

Reviewer: I think that it might be useful to talk about when you might need to "upgrade" from using `Store` to a full-blown Observable library. This isn't really clear to me; from my perspective it feels like there are just two ways to do pretty much the same thing, but Observable uses an existing concept and is more powerful, as you also get all the benefits of the rest of whichever library you're using.

While `Store` has a very compelling use-case, like the React example above shows, it feels kind of weird to me that there are two different ways to subscribe to state changes. I'd almost rather there just be one, to cut down on API surface area and remove a layer of decisions that need to be made.

Author: Observable interface was designed specifically for Observable interop. It's not really meant to be used over Store, unless you're using Observables. Internally, Observable interop using `Identity` which is the same thing as a Store. 

I'll answer the rest of this the other message with questions.

Microstates provides an easy way to convert a Microstate which represents a single value into a Observable stream of values. This is done by passing a Microstate to Observable.from function. This function will return a Observable object with a subscribe method. You can subscribe to the stream by passing an observer to the subscribe function. Once you subscribe, you will synchronously receive a microstate with middleware installed that will cause the result of transitions to be pushed through the stream.

You should be able to use to any implementation of Observables that supports Observer.from using symbol-observable. We'll use RxJS for our example.

```js

> import { from } from 'rxjs';

Reviewer: Not sure if you want to do this in the documentation, but in my own code that uses Microstates and RXJS I did this:

```javascript
import { from as createObservable } from 'rxjs';
```

just because `microstates` _also_ has a `from` function and I wanted to make it more clear which was being used at a glance.

Author: I'll do this: `import { from as observableFrom } from 'rxjs'`
